File: src/player/time_stamp.rs
use std::time::Duration;

use serde::{Deserialize, Serialize};

/// Timestamp of a source
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct TimeStamp {
    /// The current positoin
    pub current: Duration,
    /// The total length
    pub total: Duration,
}

impl TimeStamp {
    /// Creates new timestamp
    pub fn new(current: Duration, total: Duration) -> Self {
        Self { current, total }
    }
}

File: uamp-client/helper/timestamp.js
import Duration from "./duration.js";

export default class Timestamp {
    /**
     * Creates new timestamp
     * @param {Duration} current 
     * @param {Duration} total 
     */
    constructor(current, total) {
        /** @type {Duration} */
        this.current = current;
        /** @type {Duration} */
        this.total = total;
    }

    /**
     * Creates new timestamp from the given object
     * @param {*} obj 
     * @returns 
     */
    static from(obj) {
        return new Timestamp(
            Duration.from(obj.current),
            Duration.from(obj.total)
        );
    }
}
